### PR TITLE
Add --extend-exclude parameter

### LINKED ISSUE #1571: Please offer a way to exclude dirs *in addition to the default exclusions*
**Is your feature request related to a problem? Please describe.**

I needed to exclude one auto-generated file so I naively specified `--exclude thefile.py`. Unfortunately that disabled the useful standard exclusions, which I did not realize until it broke a unit test running on Jenkins.

**Describe the solution you'd like**

I would like a way to exclude one or more files or directories *in addition to the standard exclusion defaults*.

**Describe alternatives you've considered**

The simplest UI change would be a new command-line option such as `--alsoexclude` meaning "exclude the specified files and directories in addition to other exclusions." The example above would read `--alsoexclude thefile.py`.

Please also consider allowing that option to be repeated. It seems a very natural thing and it saves the hassle of generating a regex to exclude multiple items. Thus: `--alsoexclude foo.py --alsoexclude .private --alsoexclude a/path/somewhere`. 

Ideally I would suggest that same capability for `--exclude` and `--include`, because regular expressions can be hard to read and can easily have subtle bugs. But I realize that is a stretch since it changes an existing interface.

Another option is to add a boolean option that means "exclude the standard defaults", such as `--excludedefaults`. Then the example above would read `--excludedefaults --exclude thefile.py`

Yet another option is to have some magic value that means "the standard exclusion defaults". But I don't know what it could be without risking collision with files and directory names.

**Additional context**

None.

### PR DESCRIPTION
**Pull request by thejcannon**
Adding the `--extend-exclude` parameter to add additional exclude regexes on top of the `--exclude` regexes. This is especially useful if you want to specify _additional_ files/directories on top of the very useful default ones (as demonstrated by the change to our own `pyproject.toml`).

The parameter name also matches `flake8`'s parameter of `--extend-exclude`.

--- 

Fixes #1571

### CODE REVIEW

File: src/black/__init__.py [hunk 1]
```diff
@@ -602,6 +612,18 @@ def main(
     ctx.exit(report.return_code)
 
 
+def test_regex(
+    ctx: click.Context,
+    regex_name: str,
+    regex: Optional[str],
+) -> Optional[Pattern]:
```
Comment (thejcannon) [inline src/black/__init__.py:619]: I don't know the `mypy` incantation to say this returns an optional if `regex` was an optional. Suggestions welcome 👍

Comment (JelleZijlstra) [reply src/black/__init__.py:619]: You'd have to do it with overloads, but not sure that's worth bothering with. Something like 

`@overload
def test_regex(regex: str) -> Pattern: ...
@overload
def test_regex(regex: None) -> None: ...`

**Review by JelleZijlstra — APPROVED** [review]

File: src/black/__init__.py (hunk 1 above)
Comment (thejcannon) [reply src/black/__init__.py:619]: I suppose that makes sense.

> Looks like all that would gain us is removing an assert further down?

yup. Totally not worth 😄

**Comment by ichard26** [comment]
To be clear, I want to do a review and already have a comment in mind, but I unfortunately don't have time to put it in words right now, so I would appreciate holding off on the merge for the time being. Thanks!

File: tests/test_black.py [hunk 2]
```diff
@@ -1433,6 +1443,35 @@ def test_get_sources_with_stdin_filename_and_exclude(self) -> None:
...
+                force_exclude=None,
+                report=report,
+                stdin_filename=stdin_filename,
+            )
+        )
+        self.assertEqual(sorted(expected), sorted(sources))
+
+    @patch("black.find_project_root", lambda *args: THIS_DIR.resolve())
+    def test_get_sources_with_stdin_filename_and_extend_exclude(self) -> None:
+        # Extend exclude shouldn't exclude stdin_filename since it is mimicing the
```
Comment (ichard26) [inline tests/test_black.py:1456]: typo :)
```suggestion
        # Extend exclude shouldn't exclude stdin_filename since it is mimicking the
```

File: src/black/__init__.py [hunk 3]
```diff
@@ -602,6 +612,18 @@ def main(
...
+    try:
+        return re_compile_maybe_verbose(regex) if regex is not None else None
+    except re.error:
+        err(f"Invalid regular expression for {regex_name} given: {regex!r}")
+        ctx.exit(2)
```
Comment (ichard26) [inline src/black/__init__.py:624]: The more I look at this code, perhaps this would be better handled via click argument callbacks? Then, main() would either get None or a compiled regex right from the get go. IDK and IDC that much though

If that would be an improvement, it should probably be handled in another PR.

File: README.md [hunk 4]
```diff
@@ -135,11 +135,17 @@ Options:
                                   hg|\.mypy_cache|\.nox|\.tox|\.venv|\.svn|_bu
                                   ild|buck-out|build|dist)/]
 
+  --extend-exclude TEXT           Like --exclude, but adds additional files
+                                  and directories on top of the excluded
+                                  ones. (Useful if you simply want to add to
+                                  the default)
```
Comment (ichard26) [inline README.md:141]:
```suggestion
                                  ones (useful if you simply want to add to
                                  the default).
```

This style has some history, but it honestly really doesn't matter. British and American English are both used with no consistency already :p
```
  --pyi                           Format all input files like typing stubs
                                  regardless of file extension (useful when
                                  piping source on standard input).
```

*don't forget update the docs/installation_and_usage.md document too if you decide this is okay

File: README.md [hunk 5]
```diff
@@ -135,11 +135,17 @@ Options:
                                   hg|\.mypy_cache|\.nox|\.tox|\.venv|\.svn|_bu
                                   ild|buck-out|build|dist)/]
 
+  --extend-exclude TEXT           Like --exclude, but adds additional files
```
Comment (ichard26) [inline README.md:138]: This comment has nothing to do with this line, but I can't comment on the line I want to due to GitHub Review UI limitations :(

https://github.com/psf/black/pull/2005/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R160

This explanation is already not complete, notably `using configuration from [PATH]` lines are enabled by this flag, but ignoring that, I think just stating "due to exclusion configuration" or "due to exclusion regexes" would be an improvement.

*don't forget update the docs/installation_and_usage.md document too if you decide this is okay

**Review by ichard26 — COMMENTED** [review]
Approving the changes generally, but there are a few nits and minor issues that would be nice to have resolved :)

I was going to complain about how someone can pass both `--exclude` and `--extend-exclude` at the same and how it would be confusing, buuuutttt... after some more thinking, there's some valid use cases for this configuration design.

Thanks for submitting an excellent PR! Thank you for the time investment, it really helps out the project. Why not add yourself to the AUTHORS list at the end of the README?

File: README.md (hunk 4 above)
Comment (JelleZijlstra) [reply README.md:141]: Just curious, do you mean this formatting change is a British/American difference?

Comment (ichard26) [reply README.md:141]: I poorly wrote that comment (as I always do unfortunately). I meant that consistency in our docs is already a bit hazy, using the mixture British and American English in our docs as an example. And that this comment is effectively a nit.

File: src/black/__init__.py (hunk 3 above)
Comment (thejcannon) [reply src/black/__init__.py:624]: I was thinking that when making the change, but ultimately figured it was this way on purpose 😄.

I'll swing back around for this if no one beats me to it.

File: README.md (hunk 5 above)
Comment (thejcannon) [reply README.md:138]: I like the suggestion, and went with "exclusion patterns" (as it slightly mirrors the wording `pyproject.toml` doc).
And speaking of, I edited that doc to mention the other exclusion options as well (by name)

File: src/black/__init__.py (hunk 3 above)
Comment (thejcannon) [reply src/black/__init__.py:624]: I went ahead and made a PR on top of this branch. I'll switch branches when this PR is in.

https://github.com/thejcannon/black/pull/1

Comment (thejcannon) [reply src/black/__init__.py:624]: Something tells me I won't be able to switch _repos_ for that PR though. Feel free to comment on it, or wait until I republish and target this repo.

File: README.md [hunk 6]
```diff
@@ -615,7 +606,8 @@ Multiple contributions by:
 - [Jose Nazario](mailto:jose.monkey.org@gmail.com)
 - [Joseph Larson](mailto:larson.joseph@gmail.com)
 - [Josh Bode](mailto:joshbode@fastmail.com)
-- [Josh Holland](mailto:anowlcalledjosh@gmail.com)
+- [Josh Bode](mailto:joshbode@fastmail.com)
```
Comment (thejcannon) [inline README.md:609]:
```suggestion
- [Josh Bode](mailto:joshbode@fastmail.com)
- [Josh Holland](mailto:anowlcalledjosh@gmail.com)
```

Comment (thejcannon) [inline README.md:609]:
```suggestion
```

File: README.md [hunk 7]
```diff
@@ -615,7 +606,9 @@ Multiple contributions by:
 - [Jose Nazario](mailto:jose.monkey.org@gmail.com)
 - [Joseph Larson](mailto:larson.joseph@gmail.com)
 - [Josh Bode](mailto:joshbode@fastmail.com)
+- [Josh Bode](mailto:joshbode@fastmail.com)
```
Comment (thejcannon) [inline README.md]:
```suggestion
```

**Review by JelleZijlstra — APPROVED** [review]

**Comment by JelleZijlstra** [comment]
@ichard26 are you OK with merging this now? We can handle the click callbacks for regexes in a separate PR (as you said above). Thanks for your work @thejcannon!

**Review by ichard26 — APPROVED** [review]
@JelleZijlstra LGTM! Thanks for checking in with me, you didn't have to. Thanks @thejcannon for the good work!

> Look ma! I contribute to open source!

p.s. thanks for the little fun you portrayed in this commit message, it's a little light in what's a very dark day for me :)

File: src/black/__init__.py (hunk 3 above)
Comment (thejcannon) [reply src/black/__init__.py:624]: Done and dusted: https://github.com/psf/black/pull/2016
